src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -111,5 +119,106 @@ class EntityClient (requestContext: RequestContext) extends Actor {
+  private def withRequiredAttributes(entityType: String, headers: Seq[String])(op: (Map[String, String] => Unit)):Unit = {
+    ModelSchema.getRequiredAttributes(entityType) match {
+      case Failure(regret) => requestContext.complete(HttpResponse(BadRequest, regret.getMessage))
+      case Success(requiredAttributes) => {
+        if( !requiredAttributes.keySet.subsetOf(headers.toSet) ) {
+          requestContext.complete( HttpResponse(BadRequest, "TSV is missing required attributes: " + (requiredAttributes.keySet -- headers).mkString(", ")) )
+        } else {
+          op(requiredAttributes)
+        }
+      }
+    }
+  }
+
+  def addToCollectionMembersAttribute( entityType: String, entityName: String, newMembers: Seq[(String, String)], membersAttributeName: String ) = {
+    //STUB: return a partial batch Rawls call for this:
+    //(entityName, entityType).attrs[membersAttributeName].extend(newMembers)
+  }
+
+  def setAttributesOnEntity(entityType: String, entityName: String, attributesMap: Map[String, String], referenceTypes: Map[String, String] ) = {
+    //for all attributes in attributesMap
+    // if the key is present in referenceTypes, the value is an entity name and referenceTypes(key) contains that entity's type
+    //MAYBE-DO: Parse out types provided in the TSV header? Or allow the model to specify types for optionally specified attributes?
+
+    //STUB: return a batch of AddUpdateOperations
+  }
+
+  def batchCallToRawls( calls: Any ) = {
+    //STUB: stitch the batch calls together and fire off the Rawls request.
+    //requestContext.complete(...)
+  }
+
+  def addToCollectionTypeFromTSV( workspaceNamespace: String, workspaceName: String, entityType: String, memberType: String, membersAttributeName: String, tsv: TSVLoadFile ) = {
+    //group together updates to distinct entities, construct the list of entities to add references to, and poke it into whatever the "members" field is called.
+    val tsvFieldName = memberType + "_id"
+    tsv.tsvData.groupBy( _(tsv.firstColumnHeader) ).map({
+      case (entityName, rows) => addToCollectionMembersAttribute( entityType, entityName, rows.map(row => (memberType, row(tsvFieldName))), membersAttributeName )
+    })
+  }
+
+  //TSVs can be of two forms: they can either populate the "members" list of a collection entity,
+  //or they can populate the attributes on an(y kind of) entity.
+  def upsertEntitiesFromTSV(workspaceNamespace: String, workspaceName: String, tsv: TSVLoadFile): Unit = {
+    //the first column header of the tsv defines the entity type we're trying to import or update.
+    //This is magically verified in the withFoo functions below.
+    val entityType = tsv.firstColumnHeader.stripSuffix("_id")
+
+    withMemberCollectionType(entityType) { memberTypeOpt =>
+      //This magical list of conditions determines whether the TSV is populating the "members" attribute of a collection type entity.
+      if( memberTypeOpt.isDefined
```

Comment:
I find this big if hard to understand, what is the exact meaning of the cases in here?  can we just make that a function that returns boolean with a readable name instead?